### PR TITLE
storage node quick check and startup validation

### PR DESCRIPTION
**Pull request by aligeti**
On initial startup of storage node server

Following changes are made.
On First time setup as a storage node server
================================
1. The storage node server will run without putting any warnings or errors,  if available free hard drive disk space is greater than or equal to allocated storage node
2. The storage node server will exit with error "Not enough space" , if available free hard drive space is less allocated storage node

Already setup as a storage node, but restarted 
=================================
1. The storage node server will run without any warnings, or errors, if the free hard drive disk space is  greater than remaining allocated storage node space ( which is difference of Allocated Space - Total used space)
2. The storage node server will run with warning "Warning!!! Used more space than allocated" if the total used space by storage node has exceeded the allocated space, may be the last time it uploaded a pieceID just went above the allocated limits. Probability of happening this scenario is very low, as the satellite will never use this node when its space is limited. Just in case if it happens, the server will still run but will issue a warning.  
3. The storage node server will run with warning "Warning!!! Disk space is less than remaining allocated space",  if the remaining hard drive disk space is less than remaining allocated storage node space

### CODE REVIEW

File: pkg/piecestore/rpc/server/server.go [hunk 1]
```diff
@@ -53,24 +56,80 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
 // Initialize -- initializes a server struct
 func Initialize(ctx context.Context, config Config, pkey crypto.PrivateKey) (*Server, error) {
 	dbPath := filepath.Join(config.Path, "piecestore.db")
 	dataDir := filepath.Join(config.Path, "piece-store-data")
 
+	// read the allocated disk space from the config file
+	allocatedDiskSpace := config.AllocatedDiskSpace
+
+	// get the disk space details
+	diskSpace := DiskUsage("/")
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: We need to check diskusage of the drive used for `dataDir`

File: pkg/piecestore/rpc/server/server.go [hunk 2]
```diff
@@ -53,24 +56,80 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+		totalUsed = 0x00
+	}
+
+	// check your hard drive is big enough
+	// on restarting the Piece node server, assuming already been working as a node
+	if uint64(totalUsed) != 0x00 {
+		// used above the alloacated space
+		if uint64(totalUsed) >= allocatedDiskSpace {
+			log.Println("Warning!!! Used more space then allocated")
+			/** [TODO] any special handling needed here ... */
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: return here

File: pkg/piecestore/rpc/server/server.go [hunk 3]
```diff
@@ -53,24 +56,80 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+		} else {
+			// the available diskspace is less than remaining allocated space
+			if diskSpace.Free < (allocatedDiskSpace - uint64(totalUsed)) {
+				log.Println("Warning!!! Disk space is less than remaining allocated space")
+				/** [TODO] any special handling needed here ... */
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: return here

File: pkg/piecestore/rpc/server/server.go [hunk 4]
```diff
@@ -53,24 +56,80 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+	// get how much is currently used, if for the first time totalUsed = 0
+	totalUsed, err := db.SumTTLSizes()
+	if err != nil {
+		//first time setup
+		totalUsed = 0x00
+	}
+
+	// check your hard drive is big enough
+	// on restarting the Piece node server, assuming already been working as a node
+	if uint64(totalUsed) != 0x00 {
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: This `if` mountain is a little hard to follow. Flatten so all ifs are at the same level if possible to prevent weird logic for missed `if else` statements

File: pkg/piecestore/rpc/server/server.go (hunk 1 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: I changed the DiskUsage func to golang supported disk package which returns the hard drive's used, free and total and bunch of hard drive info. Seems to be very straight forward and golang supported.

Changed as suggested. Changed the function name to DirSize()

File: pkg/piecestore/rpc/server/server.go (hunk 2 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done

File: pkg/piecestore/rpc/server/server.go (hunk 3 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done

File: pkg/piecestore/rpc/server/server.go (hunk 4 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: Flattened the if's

File: pkg/piecestore/rpc/server/server.go [hunk 5]
```diff
@@ -53,24 +56,83 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+	if totalUsed >= allocatedDiskSpace {
+		log.Println("Warning!!! Used more space then allocated")
+		/** [TODO] any special handling needed here ... */
+		return &Server{DataDir: dataDir, DB: db, pkey: pkey, totalAllocated: allocatedDiskSpace}, nil
+	}
+
+	// the available diskspace is less than remaining allocated space,
+	// due to change of setting before restarting
+	if freeDiskSpace < (allocatedDiskSpace - totalUsed) {
+		log.Println("Warning!!! Disk space is less than remaining allocated space")
```
Comment (nfarah86) [inline pkg/piecestore/rpc/server/server.go]: *total nit pick*- i believe we are trying to stick . with the zap library:  https://godoc.org/go.uber.org/zap#Logger.Sugar

zap.S().Warnf("Warning!!! Disk space is less than remaining allocated space")

File: pkg/piecestore/rpc/server/server.go [hunk 6]
```diff
@@ -53,24 +56,83 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
 // Initialize -- initializes a server struct
 func Initialize(ctx context.Context, config Config, pkey crypto.PrivateKey) (*Server, error) {
 	dbPath := filepath.Join(config.Path, "piecestore.db")
 	dataDir := filepath.Join(config.Path, "piece-store-data")
 
+	// read the allocated disk space from the config file
+	allocatedDiskSpace := config.AllocatedDiskSpace
+
+	// get the disk space details
+	diskSpace, err := disk.Usage("/")
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: get drive from config.path

File: pkg/piecestore/rpc/server/server.go [hunk 7]
```diff
@@ -53,24 +56,83 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+	}
+
 	db, err := psdb.Open(ctx, dataDir, dbPath)
 	if err != nil {
 		return nil, err
 	}
 
-	return &Server{DataDir: dataDir, DB: db, pkey: pkey}, nil
+	// get how much is currently used, if for the first time totalUsed = 0
+	totalUsed, err := DirSize(dataDir)
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: use the database call instead of the file system check to get used space

File: pkg/piecestore/rpc/server/server.go [hunk 8]
```diff
@@ -53,24 +56,83 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+		/** [TODO] any special handling needed here ... */
+		return &Server{DataDir: dataDir, DB: db, pkey: pkey, totalAllocated: allocatedDiskSpace}, nil
+	}
+
+	// the available diskspace is less than remaining allocated space,
+	// due to change of setting before restarting
+	if freeDiskSpace < (allocatedDiskSpace - totalUsed) {
+		log.Println("Warning!!! Disk space is less than remaining allocated space")
+		/** [TODO] any special handling needed here ... */
+		return &Server{DataDir: dataDir, DB: db, pkey: pkey, totalAllocated: allocatedDiskSpace}, nil
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: let's set totalAllocated equal to freeDiskSpace

File: pkg/piecestore/rpc/server/server.go (hunk 5 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done

File: pkg/piecestore/rpc/server/server.go (hunk 7 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done

File: pkg/piecestore/rpc/server/server.go (hunk 8 above)
Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done... also made the change to similar situation when the user decides to allocates less than what he already used then I set the allocated space to what is already used.

File: pkg/piecestore/rpc/server/server.go [hunk 9]
```diff
@@ -53,24 +57,87 @@ func (c Config) Run(ctx context.Context, server *provider.Provider) (err error)
...
+		allocatedDiskSpace = freeDiskSpace
+		zap.S().Warnf("Disk space is less than requested allocated space, allocating = %d Bytes", allocatedDiskSpace)
+		return &Server{DataDir: dataDir, DB: db, pkey: pkey, totalAllocated: allocatedDiskSpace}, nil
+	}
+
+	// on restarting the Piece node server, assuming already been working as a node
+	// used above the alloacated space, user changed the allocation space setting
+	// before restarting
+	if totalUsed >= allocatedDiskSpace {
+		allocatedDiskSpace = totalUsed
```
Comment (aleitner) [inline pkg/piecestore/rpc/server/server.go]: We don't want to increase the allocatedDiskSpace. We shouldn't change it in this case.

Comment (aligeti) [reply pkg/piecestore/rpc/server/server.go]: done
